framework/include/cppmicroservices/detail/BundleTracker.tpp

```diff
@@ -0,0 +1,273 @@
+    US_UNUSED(l);
+    if (d->trackedBundle.Load() &&
+        !d->Tracked()->closed) { /* If BundleTracker is open */
+      return;
+    }
+
+    DIAG_LOG(*d->context.GetLogSink())
+      << "BundleTracker<T>::Open: " << d->stateMask;
+
+    t.reset(new _TrackedBundle(this, d->customizer ? d->customizer.get() : d->q_func()));
+    try {
+      // Attempt to drop old listener
+      d->context.RemoveListener(std::move(d->listenerToken));
+      // Make new listener
+      d->listenerToken = d->context.AddBundleListener(
+        std::bind(&_TrackedBundle::BundleChanged, t, std::placeholders::_1));
+
+      std::vector<Bundle> bundles = d->GetInitialBundles(d->stateMask);
+      t->SetInitial(bundles);
+    } catch (const std::invalid_argument& e) {
+      // Remove listener and rethrow
+      d->context.RemoveListener(std::move(d->listenerToken));
+      throw std::runtime_error(
+        std::string("unexpected std::invalid_argument exception: ") + e.what());
+    }
+    d->trackedBundle.Store(t);
+  }
+  t->TrackInitial();
+}
+
+template<class T>
+void BundleTracker<T>::Close()
+{
+  std::shared_ptr<_TrackedBundle> outgoing = d->trackedBundle.Load();
+  {
+    auto l = d->Lock();
+    US_UNUSED(l);
+    
+    try {
+      d->context.RemoveListener(std::move(d->listenerToken));
+    } catch (const std::runtime_error&) {
+      /* Rescue if context is stopped or invalid */
+    }
+
+    if (outgoing == nullptr) {
+      return;
+    }
+
+    // Check for already closed
+    if (d->Tracked()->closed) {
+      return;
+    }
+
+    DIAG_LOG(*d->context.GetLogSink())
+      << "BundleTracker<T>::close:" << d->stateMask;
+
+    outgoing->Close();
+    d->Modified();         /* log message */
+    outgoing->NotifyAll(); /* wake up any waiters */
+  }
+
+  try {
+    outgoing->WaitOnCustomizersToFinish();
+  } catch (const std::exception&) {
+    // Rescue CounterLatch issues
+  }
+
+  auto bundles = GetBundles();
+  for (auto& bundle : bundles) {
+    outgoing->Untrack(bundle, BundleEvent());
+  }
+}
+
+template<class T>
+std::vector<Bundle> BundleTracker<T>::GetBundles()
+{
+  std::vector<Bundle> bundles;
+  auto t = d->Tracked();
+  if (!t) { /* If BundleTracker is not open */
+    return bundles;
+  }
+  {
+    auto l = t->Lock();
+    US_UNUSED(l);
+    d->GetBundles_unlocked(bundles, t.get());
+  }
+  return bundles;
+}
+
+template<class T>
+std::optional<typename BundleTracker<T>::TrackedParamType> BundleTracker<T>::GetObject(
+  const Bundle& bundle)
+{
+  std::optional<typename BundleTracker<T>::TrackedParamType> optional;
+  auto t = d->Tracked();
+  if (!t || !bundle) { /* If BundleTracker is not open or if bundle is invalid */
+    return std::nullopt;
+  }
+  {
+    auto l = t->Lock();
+    US_UNUSED(l);
+    optional = t->GetCustomizedObject_unlocked(bundle);
+  }
+  return optional;
+}
+
+template<class T>
+typename BundleTracker<T>::TrackingMap BundleTracker<T>::GetTracked()
+{
+  BundleTracker<T>::TrackingMap map;
```

Comment:
Consider renaming the variable to something more descriptive.

Reply:
Changed to `trackingMap`